app/services/open_humans_service.rb

```diff
@@ -0,0 +1,116 @@
+    response_json = JSON.parse(response.body)
+    update_tokens(user.open_humans_profile,response_json)
+  end
+
+  def refresh_token(user)
+    oh_profile = user.open_humans_profile
+    url = URI.parse('https://www.openhumans.org/oauth2/token/')
+    req = Net::HTTP::Post.new(url.request_uri)
+    req.set_form_data({'grant_type' => 'refresh_token',
+                       'refresh_token' => oh_profile.refresh_token
+                       })
+    http = Net::HTTP.new(url.host, url.port)
+    http.use_ssl = (url.scheme == 'https')
+    response = http.request(req)
+    response_json = JSON.parse(response.body)
+    update_tokens(oh_profile,response_json)
+  end
+
+
+  def set_open_humans_ids(oh_profile)
+    uri = URI.parse('https://www.openhumans.org/api/direct-sharing/project/exchange-member/')
+    uri_params = {:access_token => oh_profile.access_token}
+    uri.query = URI.encode_www_form(uri_params)
+    res = Net::HTTP.get_response(uri)
+    res_json = JSON.parse(res.body)
+    oh_profile.project_member_id = res_json["project_member_id"]
+    oh_profile.open_humans_user_id = res_json["username"]
+    oh_profile.save
+  end
+
+  def upload_opensnp_id(user)
+    uri = URI.parse("https://www.openhumans.org/api/direct-sharing/project/files/upload/?access_token=#{user.open_humans_profile.access_token}")
+    boundary = '0P3NSNPH34RT50PENHUM4N5'
+    http = Net::HTTP.new(uri.host, uri.port)
+    http.use_ssl = (uri.scheme == 'https')
+    upload = Net::HTTP::Post.new(uri.request_uri)
+    upload.content_type = "multipart/form-data; boundary=#{boundary}"
+    post_body = generate_form_body(user,boundary)
+    upload.body = post_body.join
+    puts post_body.join
+    http.request(upload)
+  end
+
+  def delete_opensnp_id(user)
+    url = URI.parse("https://www.openhumans.org/api/direct-sharing/project/files/delete/?access_token=#{user.open_humans_profile.access_token}")
+    req = Net::HTTP::Post.new(url.request_uri)
+    oh_profile = user.open_humans_profile
+    req.set_form_data({
+                project_member_id: oh_profile.project_member_id,
+                all_files: true
+                })
+    http = Net::HTTP.new(url.host, url.port)
+    http.use_ssl = (url.scheme == 'https')
+    http.set_debug_output($stdout)
+    http.request(req)
+  end
+
+  private
+
+  def update_tokens(oh_profile,tokens)
+    oh_profile.expires_in = Time.now + tokens["expires_in"]
+    oh_profile.access_token = tokens["access_token"]
+    oh_profile.refresh_token = tokens["refresh_token"]
+    oh_profile.save
+  end
+
+  def generate_json(user)
+    json = {
+            user_name: user.name,
+            user_id: user.id,
+            has_sequence: user.has_sequence,
+            user_uri: "https://opensnp.org/users/#{user.id}"
+           }
+  end
+
+  def generate_metadata(user)
+    metadata = generate_json(user)
+    metadata['tags'] = ['opensnp']
+    metadata['description'] = "links to openSNP user #{user.id}"
+    return metadata
```

Comment:
this is actually not redundant as it otherwise breaks the hash…

Reply:
This seems to be pretty redundant, as it is the last expression in the method. Am I missing something?

Reply:
I don't know. If I'm not running the `return` it only returns `"links to openSNP user #{user.id}"` instead of the whole hash. That's why I put it back in, otherwise the OH API doesn't accept our upload.

Reply:
Aaaaah! Now I get it. 💡  You don't need the `return`, but you still need the `metadata`. Without the `metadata` in the end, it returns the return value of the assignment above it, which is the value you assign.